src/MOI_wrapper.jl

```diff
@@ -93,30 +99,445 @@ Base.@kwdef struct ConicCache
     A::SparseMatrixCSC{Float64, Int}
     b::Vector{Float64}
     c::Vector{Float64}
+    index_map::MOIU.IndexMap
+    conic_form::CONIC_FORM
 end
-
 const CACHE_TYPE = Union{
     Nothing,
     QPCache,
     ConicCache,
 }
 
+Base.@kwdef struct QPForwBackCache
+    dz::Vector{Float64}
+    dλ::Vector{Float64}
+    dν::Vector{Float64}
+end
+Base.@kwdef struct ConicForwCache
+    du::Vector{Float64}
+    dv::Vector{Float64}
+    dw::Vector{Float64}
+end
+Base.@kwdef struct ConicBackCache
+    g::Vector{Float64}
+    πz::Vector{Float64}
+end
+const CACHE_FORW_TYPE = Union{
+    Nothing,
+    QPForwBackCache,
+    ConicForwCache,
+}
+const CACHE_BACK_TYPE = Union{
+    Nothing,
+    QPForwBackCache,
+    ConicBackCache,
+}
+# TODO bechmark access to these, ^^, type unstable type
+# possibly add fields for all of them
+
+const MOIDD = MOI.Utilities.DoubleDicts
+
+# TODO: possibly use sparse vector/MatrixCSC format for better performance
+# will need to keep indices correct
+Base.@kwdef struct DiffInputCache
+    dx::Dict{VI, Float64} = Dict{VI, Float64}()# dz for QP
+    # ds
+    # dy #= [d\lambada, d\nu] for QP
+    dA::MOIDD.DoubleDict{Dict{VI,Float64}} = MOIDD.DoubleDict{Dict{VI,Float64}}() # also includes G for QPs
+    dAv::MOIDD.DoubleDict{Dict{VI,Vector{Float64}}} = MOIDD.DoubleDict{Dict{VI,Vector{Float64}}}() # also includes G for QPs
+    db::MOIDD.DoubleDict{Float64} = MOIDD.DoubleDict{Float64}() # also includes h for QPs
+    dbv::MOIDD.DoubleDict{Vector{Float64}} = MOIDD.DoubleDict{Vector{Float64}}() # also includes h for QPs
+    dc::Dict{VI, Float64} = Dict{VI, Float64}()
+    dQ::Dict{Tuple{VI, VI}, Float64} = Dict{Tuple{VI, VI}, Float64}()
+end
+
+abstract type AbstractDiffAttribute end
+Base.broadcastable(attribute::AbstractDiffAttribute) = Ref(attribute)
+
+"""
+    ForwardIn{T}
+
+A MOI.AbstractModelAttribute to set input data to forward differentiation, that
+is, problem input data.
+The input data includes:
+[`LinearObjective`](@ref), [`ConstraintConstant`](@ref),
+[`ConstraintCoefficient`](@ref) and [`QuadraticObjective`](@ref).
+The latter can only be used in linearly constrained quadratic models.
+
+```julia
+MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), x)
+```
+"""
+struct ForwardIn{T} <: AbstractDiffAttribute end
+
+"""
+    ForwardOut{T}
+
+A AbstractDiffAttribute to set input data to backward differentiation, that
```

Comment:
if it's for backward diff, why still call the attribute forward? I think I am missing the notion attached to In and Out in this context

Reply:
It is not back, it is forward indeed. The text is wrong. Thanks for finding this!

In and Out are the input differentials and output differentials.